src/core/Pipelining.ts

```diff
@@ -0,0 +1,181 @@
+ */
+
+import * as Promise from 'bluebird';
+import {assertNotNull, assertPositive} from '../Util';
+
+/**
+ * The Pipelining can be used to speed up requests. It is build on top of asynchronous
+ * requests.
+ *
+ * The main purpose of the Pipelining is to control the number of concurrent requests
+ * when using asynchronous invocations. This can be done by setting the depth using
+ * the constructor. So you could set the depth to e.g 100 and do 1000 calls. That means
+ * that at any given moment, there will only be 100 concurrent requests.
+ *
+ * It depends on the situation what the optimal depth (number of invocations in
+ * flight) should be. If it is too high, you can run into memory related problems.
+ * If it is too low, it will provide little or no performance advantage at all. In
+ * most cases a Pipelining and a few hundred asynchronous calls should not lead to any
+ * problems. For testing purposes we frequently have a Pipelining of 1000 or more
+ * concurrent requests to be able to saturate the system.
+ *
+ * A Pipelining provides its own backpressure on the system. So there will not be more
+ * in flight invocations than the depth of the Pipelining. This means that the Pipelining
+ * will work fine when backpressure on the member is disabled (default). Also
+ * when it it enabled it will work fine, but keep in mind that the number of concurrent
+ * invocations in the Pipelining could be lower than the configured number of invocation
+ * of the Pipelining because the backpressure on the member is leading.
+ *
+ * The Pipelining has been marked as Beta since we need to see how the API needs to
+ * evolve. But there is no problem using it in production. We use similar techniques
+ * to achieve high performance.
+ *
+ * @param E
+ */
+export class Pipelining<E> {
+    private readonly depth: number;
+    private readonly next: () => void;
+    private loadGenerator: () => Promise<E>;
+    private rejected: boolean = false;
+    private done: boolean = false;
+    private resolve: (thenableOrResult?: (void | E[]) | PromiseLike<void | E[]>) => void;
+    private reject: (err?: any) => void;
+    private resolvingCount: number = 0;
+    private index: number = 0;
+    private results: E[] = [];
+
+    /**
+     * Creates a Pipelining with the given depth and load generator.
+     *
+     * @param depth the maximum number of concurrent calls allowed in this Pipelining.
+     * @param loadGenerator a generator-like function that returns a promise
+     *  or null if the generator is exhausted.
+     * @param storeResults a boolean value indicating that whether the results of the requests
+     *  should be stored in the Pipelining or not. If this value is set to false, it is expected
+     *  from the user to handle the results of the requests within the load generator. It is false
+     *  by default.
+     * @throws {AssertionError} if the depth is smaller than 1.
+     * @throws {AssertionError} if the load generator is null.
```

Comment:
I think `RangeError` would be more appropriate for both throws statements. See the definition of RangeError in Nodejs documentation:
```
A subclass of Error that indicates that a provided argument was not within the set or range of acceptable values for a function; whether that is a numeric range, or outside the set of options for a given function parameter.
```

Reply:
I think the definition from node.js manual that I quoted perfectly fits the use case here. It is not limited to number intervals but it explicitly states that `whether that is a numeric range, or outside the set of options for a given function parameter`.

I looked at the article but couldn't find anything about why we shouldn't use builtin Errors. If anything, I am all for using builtin API rather than introducing new.

Reply:
I don't see how `RangeError` fits into the `if the load generator is null` case. The doc fragment that you have cited says: `a numeric range, or outside the set of options for a given function parameter`. As I see it, an argument being `null` has nothing to do with ranges or set of options, so this error would be misleading. I also tried to find `null` check cases in the Node source code, but a brief check gave me nothing.

Still, I agree that `RangeError` fits both cases much better than `AssertionError` and if we have to choose from standard errors that the one that we should use.

As for using custom error hierarchy, it has the following benefits:
* Client library users will be able to determine whether a thrown error is a library level error or it's a system level one
* Additional types of errors can be easily added, thus allowing to cover cases where standard errors don't fit in.

Reply:
What about the `TypeError` for the null check ? It is thrown when you try to access the keys of a null object. If that check does not exists, we would get `TypeError` when they call the `run` method. So, can we think this error check as a prevention for a future error and throw `TypeError` before the future error happens? I think it is a similar approach used within the Java community where `NullPointerException` is thrown when a null value is passed in to a parameter when non-null values are expected.

Reply:
+1 for `TypeError` in the `null` case.